applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_shell_element.cpp

```diff
@@ -0,0 +1,390 @@
+                break;
+            }
+            case TracedStressType::FYX:
+            {
+                direction_1 = 1;
+                direction_2 = 0;
+                stress_is_moment = false;
+                break;
+            }
+            case TracedStressType::FYY:
+            {
+                direction_1 = 1;
+                direction_2 = 1;
+                stress_is_moment = false;
+                break;
+            }
+            case TracedStressType::FYZ:
+            {
+                direction_1 = 1;
+                direction_2 = 2;
+                stress_is_moment = false;
+                break;
+            }
+            case TracedStressType::FZX:
+            {
+                direction_1 = 2;
+                direction_2 = 0;
+                stress_is_moment = false;
+                break;
+            }
+            case TracedStressType::FZY:
+            {
+                direction_1 = 2;
+                direction_2 = 1;
+                stress_is_moment = false;
+                break;
+            }
+            case TracedStressType::FZZ:
+            {
+                direction_1 = 2;
+                direction_2 = 2;
+                stress_is_moment = false;
+                break;
+            }
+            default:
+                KRATOS_ERROR << "Invalid stress type! Stress type not supported for this element!" << std::endl;
+        }
+
+        if(stress_is_moment)
+            mpPrimalElement->GetValueOnIntegrationPoints(SHELL_MOMENT_GLOBAL, stress_vector, rCurrentProcessInfo);
+        else
+            mpPrimalElement->GetValueOnIntegrationPoints(SHELL_FORCE_GLOBAL, stress_vector, rCurrentProcessInfo);
+
+        rOutput.resize(num_gps);
+        for(size_t i = 0; i < num_gps; i++)
+        {
+            rOutput(i) = stress_vector[i](direction_1, direction_2);
+        }
+
+    }
+    else
+    {
+        rOutput.resize(num_gps);
+        rOutput.clear();
+    }
+
+    KRATOS_CATCH("")
+}
+
+int AdjointFiniteDifferencingShellElement::Check(const ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY
+
+
+    GeometryType& r_geom = GetGeometry();
+
+    // verify that the variables are correctly initialized
+    KRATOS_CHECK_VARIABLE_KEY(DISPLACEMENT);
+    KRATOS_CHECK_VARIABLE_KEY(ROTATION);
+    KRATOS_CHECK_VARIABLE_KEY(VELOCITY);
+    KRATOS_CHECK_VARIABLE_KEY(ACCELERATION);
+    KRATOS_CHECK_VARIABLE_KEY(DENSITY);
+    KRATOS_CHECK_VARIABLE_KEY(SHELL_CROSS_SECTION);
+    KRATOS_CHECK_VARIABLE_KEY(THICKNESS);
+    KRATOS_CHECK_VARIABLE_KEY(CONSTITUTIVE_LAW);
+    KRATOS_CHECK_VARIABLE_KEY(ADJOINT_DISPLACEMENT);
+    KRATOS_CHECK_VARIABLE_KEY(ADJOINT_ROTATION);
+
+    // check properties
+    KRATOS_ERROR_IF(this->pGetProperties() == nullptr) << "Properties not provided for element " << this->Id() << std::endl;
+
+    const PropertiesType & props = this->GetProperties();
+
+    if(props.Has(SHELL_CROSS_SECTION)) // if the user specified a cross section ...
```

Comment:
this is legacy and will be removed in the future
I recommend not to add this

Reply:
updated